src/EventGridExtension/OutputBinding/EventGridAsyncCollector.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.EventGrid;
+using Microsoft.Azure.EventGrid.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.EventGrid
+{
+    internal sealed class EventGridAsyncCollector : IAsyncCollector<EventGridEvent>
+    {
+        // use IEventGridClient for mocking test
+        private readonly IEventGridClient _client;
+        private readonly string _topicHostname;
+        private readonly object _syncroot = new object();
+
+        private IList<EventGridEvent> _eventsToSend = new List<EventGridEvent>();
+
+        public EventGridAsyncCollector(IEventGridClient client, string topicEndpointUri)
+        {
+            _client = client;
+            _topicHostname = new Uri(topicEndpointUri).Host;
+        }
+
+        public Task AddAsync(EventGridEvent item, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            lock (_syncroot)
+            {
+                // Don't let FlushAsyc take place while we're doing this
+                _eventsToSend.Add(item);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            IList<EventGridEvent> events;
+            lock (_syncroot)
```

Comment:
nit: you probably want to minimize code in mutex block, for example, create `List<EventGridEvent>` outside and only do a reference swap inside lock section